src/hardened_only.rs

```diff
@@ -0,0 +1,135 @@
+//! Generic implementation of hardened-only key derivation.
+//!
+//! Defined in [ZIP32: Hardened-only key derivation][hkd].
+//!
+//! Any usage of the types in this module needs to have a corresponding ZIP. If you just
+//! want to derive an arbitrary key in a ZIP 32-compatible manner without ecosystem-wide
+//! coordination, use [`arbitrary::SecretKey`].
+//!
+//! [hkd]: https://zips.z.cash/zip-0032#specification-hardened-only-key-derivation
+//! [`arbitrary::SecretKey`]: crate::arbitrary::SecretKey
+
+use core::marker::PhantomData;
+
+use blake2b_simd::Params as Blake2bParams;
+use subtle::{Choice, ConstantTimeEq};
+use zcash_spec::PrfExpand;
+
+use crate::{ChainCode, ChildIndex};
+
+/// The context in which hardened-only key derivation is instantiated.
+pub trait Context {
+    /// A 16-byte domain separator used during master key generation.
+    ///
+    /// It SHOULD be disjoint from other domain separators used with BLAKE2b in Zcash
+    /// protocols.
+    const MKG_DOMAIN: [u8; 16];
+    /// The `PrfExpand` domain used during child key derivation.
+    const CKD_DOMAIN: PrfExpand<([u8; 32], [u8; 4])>;
+
+    /// The context-specific key type to which the secret key is converted at each layer.
+    ///
+    /// If no specific key type is required, this can be `[u8; 32]`.
+    type Key;
+    /// Errors that can occur when converting the secret key to the context-specific key
+    /// type, if any.
+    ///
+    /// If no specific key type is required, this can be [`std::convert::Infallible`].
+    type KeyError;
+
+    /// Converts the hardened-only secret key to the context-specific key type.
+    fn convert_to_key(sk: [u8; 32]) -> Result<Self::Key, Self::KeyError>;
+    /// Exposes the bytes of the context-specific key type.
+    fn key_bytes(sk: &Self::Key) -> &[u8; 32];
```

Comment:
Another nit for this approach is that there's no guarantee that `C::key_bytes` returns the same bytes as were passed to `C::convert_to_key` (it entirely depends on the `Context` impl). If we want to keep `C::Key` while avoiding this issue then we'd need to store both the plain bytes and the converted bytes in `HardenedOnlyKey`; if instead we remove the conversion from the main derivation then we only store the bytes once and don't have this problem.